### PR TITLE
Line numbers

### PR DESCRIPTION
**Pull request by rlisahuang**
Line numbers are back. Also fixed some CSS issues with code styling.
Screen Shot 2022-08-09 at 4 57 24 PM
Screen Shot 2022-08-09 at 4 57 07 PM

Now showing line numbers for all blocks rendered by CustomEditor by default. Adding logic around it.

### CODE REVIEW

**Comment by rlisahuang** [comment]
TODO: add more lines to the playground

**Comment by pelikhan** [comment]
Are you reusing branches? The history on this PR contains a lot of unrelated commits.

**Comment by pelikhan** [comment]
Don't add TODOs to existing PRs, move them to a new PR instead.

**Comment by rlisahuang** [comment]
This is ready to be merged for now. Will create a separate PR for extra lines in playground.

File: website/src/components/TutorialComponents/CodeBlock.tsx [hunk 1]
```diff
@@ -32,6 +34,24 @@ export function GithubDiscussionBtn({ repo }) {
     );
 }
 
+export function ResetBtn({ resetCode, input }) {
```
Comment (pelikhan) [inline website/src/components/TutorialComponents/CodeBlock.tsx:37]: this looks unrelated to line numbers

**Review by pelikhan — CHANGES_REQUESTED** [review]
Try to keep pull requests as small as possible for review. This PR contains a lot of unrelated changes to "line numbers". (don't reuse branches)

File: website/src/components/TutorialComponents/CodeBlock.tsx [hunk 2]
```diff
@@ -76,37 +96,49 @@ const CodeEditor = (props) => {
...
                             </div>
-                        ))}
-                    </pre>
+                        </pre>
+                    </div>
                 )}
             </Highlight>
             <div className={codeBlockContentStyles.buttonGroup}>
                 <CopyButton className={codeBlockContentStyles.codeButton} code={code} />
+                {props.readonly ? <></> : <ResetBtn resetCode={setCode} input={props.code} />}
```
Comment (pelikhan) [inline website/src/components/TutorialComponents/CodeBlock.tsx:141]: unrelated to line numbers?

File: website/src/components/TutorialComponents/CodeBlock.tsx [hunk 3]
```diff
@@ -20,6 +21,7 @@ export function GithubDiscussionBtn({ repo }) {
         <button
             type="button"
             aria-label="Go to GitHub discussion"
+            title="Go to GitHub discussion"
```
Comment (pelikhan) [inline website/src/components/TutorialComponents/CodeBlock.tsx:24]: unrelated to line numbers?

File: website/src/components/TutorialComponents/CodeBlock.tsx [hunk 4]
```diff
@@ -76,37 +96,49 @@ const CodeEditor = (props) => {
...
-                                    .filter((token) => !token.empty)
-                                    .map((token, tokenIndex) => (
-                                        // eslint-disable-next-line react/jsx-key
-                                        <span
-                                            {...getTokenProps({ token, key: `token-${tokenIndex}` })}
-                                        />
-                                    ))}
-                                {"\n"}
+                    <div style={{ display: "flex" }}>
+                        {props.showLineNumbers ?
```
Comment (pelikhan) [inline website/src/components/TutorialComponents/CodeBlock.tsx:100]: nit: destructure props arugments at the start of the component
nit: you can also do 

    {props.showLineNumbers && <div ....></div>}

File: website/src/components/TutorialComponents/CodeBlock.tsx [hunk 5]
```diff
@@ -126,6 +158,8 @@ CodeEditor.propTypes = {
     prism: PropTypes.object,
```
Comment (pelikhan) [inline website/src/components/TutorialComponents/CodeBlock.tsx:158]: perks of using typescript: use inline types in the function signature

File: website/src/components/TutorialComponents/index.tsx [hunk 6]
```diff
@@ -117,7 +120,7 @@ function CustomCodeEditor(props: MyProps) {
 }
 
 export default function CustomCodeBlock({ input }) {
```
Comment (pelikhan) [inline website/src/components/TutorialComponents/index.tsx:122]: add typing

File: website/src/remark/render-code-blocks.mjs [hunk 7]
```diff
@@ -171,12 +171,18 @@ export default function plugin() {
             const skipErr = meta && meta.match(skipRegex) !== null;
             const editableRegex = new RegExp(/(always-editable)/g);
             const alwaysEditable = meta && meta.match(editableRegex) !== null;
+            const lineNumRegex = new RegExp(/(showLineNumbers)/g);
```
Comment (pelikhan) [inline website/src/remark/render-code-blocks.mjs:174]: or
    const lineNumRegex = /(showLineNumbers)/g

Comment (pelikhan) [reply website/src/remark/render-code-blocks.mjs:174]: maybe `/ig` to be case insensitive

File: website/src/remark/render-code-blocks.mjs [hunk 8]
```diff
@@ -171,12 +171,18 @@ export default function plugin() {
...
 
             for (const langConfig of languageConfig.languages) {
 
                 const label = langConfig.label;
                 const highlight = langConfig.highlight;
 
+                // line numbers can be shown for all blocks through `language.config.js`,
+                // or for a specific block through `showLineNumbers`
+                // e.g. ```z3 showLineNumbers
+                const showLineNumbers = langConfig.showLineNumbers || meta && meta.match(lineNumRegex) !== null;
```
Comment (pelikhan) [inline website/src/remark/render-code-blocks.mjs:184]: use `?.` operator to call a potentially nully instance method

    meta && meta.match(lineNumRegex)

to

    meta?.match(lineNumRegex)

File: website/src/remark/render-code-blocks.mjs [hunk 9]
```diff
@@ -186,37 +192,25 @@ export default function plugin() {
                 if (!langConfig.buildConfig) {
                     // there is no runtime configured,
                     // so just add the syntax highlighting and github discussion button (if configured)
-
```
Comment (pelikhan) [inline website/src/remark/render-code-blocks.mjs:189]: looks unrelated?

File: website/src/components/TutorialComponents/CodeBlock.tsx (hunk 4 above)
Comment (rlisahuang) [reply website/src/components/TutorialComponents/CodeBlock.tsx:100]: Resolved in #98

File: website/src/components/TutorialComponents/CodeBlock.tsx (hunk 5 above)
Comment (rlisahuang) [reply website/src/components/TutorialComponents/CodeBlock.tsx:158]: resolved in #98

File: website/src/components/TutorialComponents/index.tsx (hunk 6 above)
Comment (rlisahuang) [reply website/src/components/TutorialComponents/index.tsx:122]: resolved in #98

File: website/src/remark/render-code-blocks.mjs (hunk 8 above)
Comment (rlisahuang) [reply website/src/remark/render-code-blocks.mjs:184]: resolved in #98

File: website/src/remark/render-code-blocks.mjs (hunk 7 above)
Comment (rlisahuang) [reply website/src/remark/render-code-blocks.mjs:174]: resolved in #98

**Comment by rlisahuang** [comment]
Future reviews moved to #96
